app/src/androidTest/java/org/mozilla/fenix/FirefoxTestRunner.kt

```diff
@@ -4,13 +4,81 @@
 
 package org.mozilla.fenix
 
-import android.app.Application
-import android.content.Context
+import androidx.annotation.VisibleForTesting
+import androidx.test.platform.app.InstrumentationRegistry
 import androidx.test.runner.AndroidJUnitRunner
+import okhttp3.mockwebserver.Dispatcher
+import okhttp3.mockwebserver.MockResponse
+import okhttp3.mockwebserver.MockWebServer
+import okhttp3.mockwebserver.RecordedRequest
+
+/**
+ * Create a mock webserver that accepts all requests and replies with "OK".
+ * @return a [MockWebServer] instance
+ */
+private fun createMockWebServer(): MockWebServer {
+    val server = MockWebServer()
+    server.setDispatcher(object : Dispatcher() {
+        override fun dispatch(request: RecordedRequest): MockResponse {
+            return MockResponse().setBody("OK")
+        }
+    })
+    return server
+}
+
+/**
+ * Returns the currently active instance of the ping server.
+ *
+ * @return the active [MockWebServer] instance
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.NONE)
+internal fun getPingServer(): MockWebServer {
+    val testRunner: FirefoxTestRunner = InstrumentationRegistry.getInstrumentation() as FirefoxTestRunner
+    return testRunner.pingServer
+}
+
+/**
+ * Returns the port the local ping server is listening to.
+ *
+ * @return an `Int` containing the server port.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.NONE)
+internal fun getPingServerPort(): Int {
+    val testRunner: FirefoxTestRunner = InstrumentationRegistry.getInstrumentation() as FirefoxTestRunner
+    return testRunner.pingServerPort!!
+}
 
 class FirefoxTestRunner : AndroidJUnitRunner() {
```

Comment:
@rpappalax Do you have any concerns with adding this code here?

Reply:
You can add this here but then the lazy ping server will be running in the background for every test (which it prob doesn't need to be). It seems like it might make more sense to extend the existing Mock webserver with anything you need to add and then init it in your test itself.  That way you could also shut it down right after the test was done.

Reply:
Something like this: https://github.com/mozilla-mobile/fenix/blob/master/app/src/androidTest/java/org/mozilla/fenix/ui/NavigationToolbarTest.kt#L38-L49

Reply:
+1 on everything @rpappalax said, with extra emphasis that this approach could leak state between tests if we're not super careful.  I'd sleep better if we were reinstantiating this between tests.

Reply:
The reason why I didn't do that, is that the server address needs to be know when using [the test rule](https://github.com/mozilla-mobile/fenix/pull/7639/files#diff-cc48215f38c48b0b4bfa4f9c1a9833acR39). That test rule ensures that the Glean singleton points to the local server.

Maybe there's another way to do that, but I'm not super familiar with JUnit and rules. Any suggestion on how to achieve that, @rpappalax or @Baron-Severin ?

Reply:
Apologies in advance if I'm misunderstanding your problem, but these changes _seem_ to work for me.

I'm creating a `MockWebServer` at the top of `BaselinePingTest.kt` and passing its port into the rule.  LMK if I'm missing something here and I'll try to help.

https://github.com/Baron-Severin/fenix/compare/a799d1a493d6ef79a491ad879e1e7580354e43b0...Baron-Severin:integration_test

Reply:
Woha, @Baron-Severin , you were indeed right. This works like a charm, I've updated my PR with your changes.